filter-kalman/kalman.cpp

```diff
@@ -84,54 +223,72 @@ void FTNoIR_Filter::filter(const double* input, double *output)
         return;
     }
 
+    // Note this is a terrible way to detect when there is a new
+    // frame of tracker input, but it is the best we have.
+    bool new_input = input.cwiseNotEqual(last_input).any();
+
     // Get the time in seconds since last run and restart the timer.
     const double dt = timer.elapsed_seconds();
+    dt_since_last_input += dt;
     timer.start();
 
-    // Note this is a terrible way to detect when there is a new
-    // frame of tracker input, but it is the best we have.
-    bool new_input = false;
-    for (int i = 0; i < 6 && !new_input; i++)
-        new_input = (input[i] != last_input[i]);
-
-    // Update the transitionMatrix and processNoiseCov for dt.
-    double accel_variance = accel_stddev * accel_stddev;
-    double a = dt * dt * accel_variance;  // dt^2 * accel_variance.
-    double b = 0.5 * a * dt;  // (dt^3)/2 * accel_variance.
-    double c = 0.5 * b * dt;  // (dt^4)/4 * accel_variance.
-    for (int i = 0; i < 6; i++) {
-        kalman.transitionMatrix.at<double>(i,i+6) = dt;
-        kalman.processNoiseCov.at<double>(i,i) = c;
-        kalman.processNoiseCov.at<double>(i+6,i+6) = a;
-        kalman.processNoiseCov.at<double>(i,i+6) = b;
-        kalman.processNoiseCov.at<double>(i+6,i) = b;
-    }
-    // Get the updated predicted position.
-    cv::Mat next_output = kalman.predict();
-    // If we have new tracker input, get the corrected position.
-    if (new_input) {
-        cv::Mat measurement(6, 1, CV_64F);
-        for (int i = 0; i < 6; i++) {
-            measurement.at<double>(i) = input[i];
-            // Save last_input for detecting new tracker input.
-            last_input[i] = input[i];
-        }
-        next_output = kalman.correct(measurement);
+    output = do_kalman_filter(input, dt, new_input);
+    
+    {
+        // Compute deadzone size base on estimated state variance.
+        // Given a constant input plus noise, KF should converge to the true (constant) input.
+        // This works indeed. That is the output pose becomes very still afte some time.
+        // At this point the estimated cov should be minimal. We can use this to
+        // calculate the size of the deadzone, so that in the stationary state the
+        // deadzone size is zero. Thus the tracking error due to the dz-filter 
+        // becomes zero.
+        PoseVector variance = kf.state_cov.diagonal().head(6);
```

Comment:
I should add that this only works for the adaptive KF where the discrepancy between measured position and predicted position causes the process noise cov to increase.
In normal KF state cov is constant.